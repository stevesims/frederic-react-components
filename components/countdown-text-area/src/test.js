import React from 'react';
import { mount, shallow } from 'enzyme';
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';
import { ERROR_COLOUR } from 'govuk-colours';

import CountdownTextArea from '.';

expect.extend(createMatchers(emotion));

const onChange = jest.fn();

describe('CountdownTextArea', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<CountdownTextArea />);
    expect(wrapper.exists()).toBe(true);
  });

  describe('displays a countdown based on maxLength prop', () => {
    it('contains countdown', () => {
      const wrapper = shallow(<CountdownTextArea maxLength={250} />);
      expect(wrapper.find('StyledCountdown').childAt(0).text()).toBe('250');
    });

    it('reduces countdown', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={250} onChange={onChange} />);
      expect(wrapper.find('StyledCountdown').childAt(0).text()).toBe('246');
    });
  });

  describe('displays a countdown container prop', () => {
    it('correctly shows container name attribute', () => {
      const wrapper = shallow(<CountdownTextArea name="countdown" />);
      expect(wrapper.prop('name')).toEqual('countdown--container');
    });

    it('correctly shows countdown wrapper attribute', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" name="countdown" maxLength={202} />);
      expect(wrapper.find('StyledCountdownContainer').prop('name')).toEqual('countdown--count_wrapper');
    });

    it('correctly shows countdown name attribute', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" name="countdown" maxLength={202} />);
      expect(wrapper.find('StyledCountdown').prop('name')).toEqual('countdown--count');
    });
  });

  describe('displays count as expected', () => {
    it('correctly shows 198 when 198 remaining', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={202} onChange={onChange} />);
      expect(wrapper.find('StyledCountdown').childAt(0).text()).toBe('198');
    });

    it('correctly shows -2 as count when 2 over', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={2} onChange={onChange} />);
      expect(wrapper.find('StyledCountdown').childAt(0).text()).toBe('-2');
    });
  });

  describe('sets styles as expected', () => {
    it('sets color and font-weight when overflowing', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={2} onChange={onChange} />);
      expect(wrapper).toHaveStyleRule('color', ERROR_COLOUR);
      expect(wrapper).toHaveStyleRule('font-weight', '800');
    });

    it('does not set color and font-weight when not overflowing', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={5} onChange={onChange} />);
      expect(wrapper).not.toHaveStyleRule('color', ERROR_COLOUR);
      expect(wrapper).not.toHaveStyleRule('font-weight', '800');
    });
  });

  describe('manages maxLength prop as expected', () => {
    it('sets maxLength prop if sent', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={10} onChange={onChange} />);
      expect(wrapper.find('StyledTextAreaField').prop('maxLength')).toBe(10);
    });

    it('does not set maxLength prop if not sent', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={2} noMaxLengthAttr onChange={onChange} />);
      expect(wrapper.find('StyledTextAreaField').prop('maxLength')).toBe(null);
    });

    it('maxLengths to zero when positiveOnly set', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" maxLength={2} positiveOnly onChange={onChange} />);
      expect(wrapper.find('StyledCountdown').childAt(0).text()).toBe('0');
    });

    it('omits countdown field when no maxLength given', () => {
      const wrapper = shallow(<CountdownTextArea value="Test" onChange={onChange} />);
      expect(wrapper.find('StyledCountdown').exists()).toBe(false);
    });
  });

  it('matches snapshot', () => {
    const wrapper = mount(<CountdownTextArea value="Test" maxLength={2} onChange={onChange} />);
    expect(wrapper).toMatchSnapshot();
  });
});
