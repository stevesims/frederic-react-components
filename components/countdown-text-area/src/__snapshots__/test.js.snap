// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownTextArea matches snapshot 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  color: #b10e1e;
  font-weight: 800;
}

.emotion-0 {
  box-sizing: border-box;
  font-family: "nta",Arial,sans-serif;
  font-weight: 400;
  text-transform: none;
  font-size: 16px;
  line-height: 1.25;
  width: 100%;
  padding: 5px 4px 4px;
  border: 2px solid #0b0c0c;
  min-height: 250px;
}

@media only screen and (min-width:641px) {
  .emotion-0 {
    font-size: 19px;
    line-height: 1.3;
    width: 75%;
  }
}

.emotion-0:focus {
  outline: 3px solid #ffbf47;
  outline-offset: 0;
}

.emotion-4 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  line-height: 1;
  position: relative;
  width: 0;
}

.emotion-2 {
  position: absolute;
  bottom: 12px;
  right: 10px;
  font-family: "nta",Arial,sans-serif;
}

<CountdownTextArea
  maxLength={2}
  noMaxLengthAttr={false}
  onChange={[MockFunction]}
  positiveOnly={false}
  value="Test"
>
  <StyledTextArea
    maxLength={2}
    name="CountdownTextArea--container"
    noMaxLengthAttr={false}
    onChange={[MockFunction]}
    over={true}
    positiveOnly={false}
    value="Test"
  >
    <div
      className="emotion-6 emotion-7"
      maxLength={2}
      name="CountdownTextArea--container"
      onChange={[MockFunction]}
      value="Test"
    >
      <StyledTextAreaField
        maxLength={2}
        onChange={[MockFunction]}
        value="Test"
      >
        <textarea
          className="emotion-0 emotion-1"
          maxLength={2}
          onChange={[MockFunction]}
          value="Test"
        />
      </StyledTextAreaField>
      <StyledCountdownContainer
        name="CountdownTextArea--count_wrapper"
      >
        <div
          className="emotion-4 emotion-5"
          name="CountdownTextArea--count_wrapper"
        >
          <StyledCountdown
            name="CountdownTextArea--count"
          >
            <div
              className="emotion-2 emotion-3"
              name="CountdownTextArea--count"
            >
              -2
            </div>
          </StyledCountdown>
        </div>
      </StyledCountdownContainer>
    </div>
  </StyledTextArea>
</CountdownTextArea>
`;
